source/agora/common/BitMask.d

```diff
@@ -0,0 +1,189 @@
+/*******************************************************************************
+
+    This is an implementation of a bitmask that once initialized will have a
+    fixed number of bits for holding `true` / `false` (`1` / `0`) values.
+    It allocates the required number of `ubytes` in the constructor and does not
+    allow reading or writing to bits beyond the fixed count which is set during
+    construction.
+    This type is created for use as the validators signing bitmask and any
+    changes should ensure that it does not compromise that use.
+
+*******************************************************************************/
+
+module agora.common.BitMask;
+
+public struct BitMask
+{
+    import std.algorithm;
+    import std.conv;
+    import std.format;
+    import std.range;
+
+    @safe:
+
+    /// Serialization
+    public void toString (scope void delegate (scope const char[]) @safe sink) const
+    {
+        formattedWrite(sink, "%s", iota(this.length).map!(i => this[i] ? "1" : "0").join(""));
+    }
+
+    /// Support for Vibe.d serialization
+    public string toString () const
+    {
+        string ret;
+        scope void delegate (scope const char[]) @safe sink = (scope v) { ret ~= v; };
+        this.toString(sink);
+        return ret;
+    }
+
+    pure:
+
+    /// Support for Vibe.d deserialization
+    public static BitMask fromString (scope const(char)[] str)
+    {
+        auto bitmask = BitMask(str.length);
+        str.enumerate.each!((i, c)
+        {
+            if (c == '1')
+                bitmask[i] = true;
+        });
+        return bitmask;
+    }
+
+    nothrow:
+
+    /// Count of active validators who are expected / allowed to sign the block
+    public size_t length;
```

Comment:
This should be `private`.

Reply:
Updated and added public `count` function.